glommio/src/channels/sharding.rs

```diff
@@ -0,0 +1,177 @@
+use std::cell::RefCell;
+use std::fmt::{self, Debug, Formatter};
+use std::rc::Rc;
+
+use futures_lite::{Stream, StreamExt};
+
+use crate::channels::channel_mesh::{FullMesh, Senders};
+use crate::task::JoinHandle;
+use crate::{GlommioError, Local, ResourceType, Result};
+
+/// Trait for handling sharded messages
+///
+/// * `msg` - The message to handle.
+/// * `src_shard` - Id of the shard where the msg is from.
+/// * `cur_shard` - Id of the local shard.
+pub trait Handler<T>: Clone {
+    fn handle(&self, msg: T, src_shard: usize, cur_shard: usize);
```

Comment:
Doesn't this have to be async?
The way this is now works well for integers, but it shouldn't be hard to come up with an example where you want to do some I/O processing in this handler.

Spawning a task results in a memory allocation (although using async-trait does too... oh well)

Reply:
Either spawning a task or using async-trait is expensive for a single message, I'm afraid. The idiomatic way, IMO, would be unbounded local channels with sync `send` method.

Reply:
This is still the main contention point in my view.

I don't see how we can realistically not support async function in the receiver end. The biggest problem is that there is no way to do backpressure with this.

Imagine there is a sender that is 2x as fast as the receiver: how does the receiver throttles the sender? The receiver gets the messages, puts into an unbounded local channel, but the messages keep coming and the channel will eventually run out of memory.

Since the handler consumes like this:

```
                while let Some(msg) = stream.recv().await { <== stream is good to send a new request
                    handler.handle(msg, src_shard, cur_shard)  <== request ends up using more memory in the local channel anyway, but no backpressure
                }
```

I am inclined to think we should just use `async trait` here. It is an allocation, but pushing things into a local channel may allocate memory as well.

For very high throughput use cases where the cost of the allocation is prohibitive batching and sending a vector through the channel seems like a better strategy
(you can't send vectors now, but we already agreed we'll fix it, so let's assume it will be fixed)